pkg/token/token.go

```diff
@@ -15,6 +64,81 @@ func Random(size int) (string, error) {
 	return hex.EncodeToString(token), err
 }
 
-func Name(clusterName string) string {
+// name returns the name of the token secret, computed by convention using the name of the cluster.
+func name(clusterName string) string {
 	return fmt.Sprintf("%s-token", clusterName)
 }
+
+func getSecret(ctx context.Context, ctrlclient client.Client, clusterKey client.ObjectKey) (*corev1.Secret, error) {
+	s := &corev1.Secret{}
+	key := client.ObjectKey{
+		Name:      name(clusterKey.Name),
+		Namespace: clusterKey.Namespace,
+	}
+	if err := ctrlclient.Get(ctx, key, s); err != nil {
+		return nil, err
+	}
+
+	return s, nil
+}
+
+func generateAndStore(ctx context.Context, ctrlclient client.Client, clusterKey client.ObjectKey, owner client.Object) (*string, error) {
+	tokn, err := randomB64(16)
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate token: %v", err)
+	}
+
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name(clusterKey.Name),
+			Namespace: clusterKey.Namespace,
+			Labels: map[string]string{
+				clusterv1.ClusterNameLabel: clusterKey.Name,
+			},
+		},
+		Data: map[string][]byte{
+			"value": []byte(tokn),
+		},
+		Type: clusterv1.ClusterSecretType,
+	}
+
+	//nolint:errcheck
+	controllerutil.SetControllerReference(owner, secret, ctrlclient.Scheme())
+
+	// as secret creation and scope.Config status patch are not atomic operations
+	// it is possible that secret creation happens but the config.Status patches are not applied
+	if err := ctrlclient.Create(ctx, secret); err != nil {
+		return nil, fmt.Errorf("failed to store token: %v", err)
+	}
+
+	return &tokn, nil
+}
+
+// upsertControllerRef takes controllee and controller objects, either replaces the existing controller ref
+// if one exists or appends the new controller ref if one does not exist, and returns the updated controllee
+// This is meant to be used in place of controllerutil.SetControllerReference(...), which would throw an error
+// if there were already an existing controller ref.
+func upsertControllerRef(controllee client.Object, controller client.Object) {
+	newControllerRef := metav1.NewControllerRef(controller, controller.GetObjectKind().GroupVersionKind())
+
+	// Iterate through existing owner references
+	var updatedOwnerReferences []metav1.OwnerReference
+	var controllerRefUpdated bool
+	for _, ownerRef := range controllee.GetOwnerReferences() {
+		// Identify and replace the controlling owner reference
```

Comment:
this seems not right?

Reply:
Can you clarify what you mean here by "right"--a potential correctness issue, or a style issue?

The goal of this function is to replace an existing controlling owner reference if one exists, hence its verbosity. I'd love to clean it up if you have any ideas.

Reply:
Maybe using https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/controller/controllerutil#SetOwnerReference ? But I'm not sure that covers this specific use case...

Reply:
hi @cannonpalms ,  sorry, i should describe my thought clearer.

consider the case where ownerReferences has 1 controller ref, but not kcp, then this func will add a new controller ref to kcp, but this causes problem, because only 1 controller ref is allowed.

consider another case where ownerReferences has several ref, and one of it is controller ref to kcp. then this func will remove all existing ref and add same number of duplicate controller ref to kcp.

Reply:
Ah, I see. The answer to both of your questions is actually the same.

-----------------------
**tl;dr--** What we need to do in the case of existing secrets is to replace any existing controller ref with a KThreesControlPlane controller ref while maintaining any existing non-controlling owner refs. That's exactly what this `upsertControllerRef(...)` function does; it properly handles both cases you mentioned, @mogliang.

---------------------

I evaluated `controllerutil.SetControllerReference(...)`, but this function throws an error if there is already an existing controller reference set.

From the [doc](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/controller/controllerutil#SetControllerReference) @luisdavim linked: 
```go
// Since only one OwnerReference can be a controller, it returns an error if
// there is another OwnerReference with Controller flag set.
func SetControllerReference(owner, controlled metav1.Object, scheme *runtime.Scheme) error {
```
That makes it unsuitable for our use case.

----------------------------

Knowing that we cannot use `controllerutil.SetControllerReference`, I looked at other options. I considered a naive implementation that would simply overwrite the entire slice of owner references, replacing it with a slice containing only our new controller ref for KThreesControlPlane. 
```go
controllerRef := metav1.NewControllerRef(owner, owner.GetObjectKind().GroupVersionKind())
secret.OwnerReferences = []metav1.OwnerReference{controllerRef}
```

This solution is also not an option, because it is not acceptable to remove the owner references set by other operators or humans.

----------------------------

Therefore, we needed a solution that would respect existing owner references, but remove the existing controller ref and replace it with our new KThreesControlPlane controller ref.

`controllerutil` has exactly the function we need, but it sadly is not exported:
https://github.com/kubernetes-sigs/controller-runtime/blob/2154ffbc22e26ffd8c3b713927f0df2fa40841f2/pkg/controller/controllerutil/controllerutil.go#L177-L185

Therefore, I wrote our own implementation.
```go
// upsertControllerRef takes controllee and controller objects, either replaces the existing controller ref
// if one exists or appends the new controller ref if one does not exist, and returns the updated controllee
// This is meant to be used in place of controllerutil.SetControllerReference(...), which would throw an error
// if there were already an existing controller ref.
func upsertControllerRef(controllee client.Object, controller client.Object) {
	newControllerRef := metav1.NewControllerRef(controller, controller.GetObjectKind().GroupVersionKind())
	// Iterate through existing owner references
	var updatedOwnerReferences []metav1.OwnerReference
	var controllerRefUpdated bool
	for _, ownerRef := range controllee.GetOwnerReferences() {
		// Identify and replace the controlling owner reference
		if metav1.IsControlledBy(controllee, controller) {
			updatedOwnerReferences = append(updatedOwnerReferences, *newControllerRef)
			controllerRefUpdated = true
		} else {
			// Keep non-controlling owner references intact
			updatedOwnerReferences = append(updatedOwnerReferences, ownerRef)
		}
	}
	// If the controlling owner reference was not found, add the new one
	if !controllerRefUpdated {
		updatedOwnerReferences = append(updatedOwnerReferences, *newControllerRef)
	}
	controllee.SetOwnerReferences(updatedOwnerReferences)
}
```

This function iterates over all existing owner references and builds a new slice of owner references that (1) _does not include_ any existing controller ref and (2) _does include_ our new controller ref. Exactly what we need.

Reply:
It may also be worth noting that the reason that we _can_ (and [do](https://github.com/cannonpalms/cluster-api-k3s/blob/2ec687ef14edafcc19c588a3228688eb817666df/pkg/token/token.go#L107)) use `controllerutil.SetControllerReference(...)` when generating the token for the first time and creating the secret is that, because the secret is new, we have a guarantee that there is no pre-existing controller reference nor are there other non-controlling owner references. We're working with a clean slate in `generateAndStore(...)`.

Reply:
How about I add some tests to demonstrate this?

Reply:
> We can't use `SetControllerReference` and even if we use a custom function for setting the controller reference it will return the same error if more than one is marked as the controller, for this case we don't need it to be the controller right?
> 
> ```go
> // SetOwnerReference is a helper method to make sure the given object contains an object reference to the object provided.
> // This allows you to declare that owner has a dependency on the object without specifying it as a controller.
> // If a reference to the same object already exists, it'll be overwritten with the newly provided version.
> func SetOwnerReference(owner, object metav1.Object, scheme *runtime.Scheme) error {
> ```

yeah, i think we don't need add it as controller ref, normal owner ref shall work as well.

Reply:
We do need and want to replace the _controlling_ ownership by a KThreesConfig with _controlling_ ownership by our _KThreesControlPlane_.

- After this change is deployed, we do not want existing Secrets to maintain any owner reference, controlling or non-controlling, to a KThreesConfig object
- "", we _do_ want existing Secrets to have a controlling owner reference to our KThreesControlPlane

----------------

I've just pushed additional tests for this function and `Reconcile(...)` that helped me find a bug in `upsertControllerRef(...)`, so I very much appreciate this feedback!

Reply:
I hope that I have been clear in my responses here. I'm going to resolve this thread but please resolve it if you have further concerns based on the following summary:

- We do not want to maintain the existing reference to a KThreesConfig. It must be removed. Keeping it would send the wrong message about our intentions and desired garbage collection behavior.
- We need to use a controlling ownership reference to the KThreesControlPlane. Controlling ownership is an oft-misused (or underused) concept, but is necessary in order to clearly call out that the KThreesControlPlane is responsible for managing this k3s token Secret. Although controlling ownership does not impact **default** (without a finalizer) garbage collection behavior in a way different to _non-controlling_ ownership, it is commonly used to signal that the given owner is responsible for the deletion of the child (in addition to its creation/management over time).

Reply:
Thanks cannonpalms, i just thought set owner can be easier to implement. Since upsertControllerRef has been fixed, and with good test coverage, i will perfer keep this way, and we can reuse it.